server/security/collections.js

```diff
@@ -46,7 +46,9 @@ export default function () {
     allow(type, arg, userId) {
       if (!arg) throw new Error("ifHasRole security rule method requires an argument");
       if (arg.role) {
-        return Roles.userIsInRole(userId, arg.role, Reaction.getShopId());
+        // userId is passed to getShopId to ensure that it returns the correct shop based on the User Preference
+        // if not passed, getShopId can default to primaryShop
+        return Roles.userIsInRole(userId, arg.role, Reaction.getShopId(userId));
```

Comment:
Adding a separate security review here because this code directly affects allow/deny methods:

This update from using the default `getShopId` to passing in the `userId` changes the shop that the passed in role (`arg.role`) is being checked for. This results in the user's active shop being checked against in cases where that shop is different than the default shop.

Any place where we're `allow`ing  db operations from the client has potential to be a security hole, so it's imperative that we're checking both that `a user has the correct role` - which this is doing for the active shop, _and_ that the document being updated belongs to the shop that we are checking the role of the user for. As long as those two things are in alignment, we shouldn't see any security vulnerabilities such as permitting a user with a role for `shopA` to affect a document that belongs to `shopB`.

In this PR, we're changing `ifFileBelongsToShop` as well as `ifHasRoleForActiveShop`, but we're not changing `ifShopIdMatches` or `ifShopIdMatchesThisId` 

My only concern here would be that somehow we end up checking the user roles against `shopA` and checking that a document belongs to a different shop, `shopB` which would permit a user to perform an operation against a file they don't own.

I believe that we should update `ifShopIdMatches` and `ifShopIdMatchesThisId` to have the `userId` explicitly passed as well, even though they are likely getting the correct shopId anyway.

Reply:
Thanks. I'll get back to this shortly asap